ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/KsqlResource.java

```diff
@@ -152,234 +179,216 @@ public KsqlResource(
 
-      try {
-        result.add(executeStatement(statementText, statement, streamsProperties));
-      } catch (final Exception e) {
-        return Errors.serverErrorForStatement(e, statementText, result);
-      }
+      return executeStatements(statements, propertyOverrides);
+    } catch (final KsqlRestException e) {
+      return e.getResponse();
+    } catch (final KsqlStatementException e) {
+      return Errors.badStatement(e, e.getSqlStatement());
+    } catch (final KsqlException e) {
+      return Errors.badRequest(e);
+    } catch (final Exception e) {
+      return Errors.serverErrorForStatement(e, request.getKsql(), new KsqlEntityList());
     }
-    return Response.ok(result).build();
   }
 
-  // CHECKSTYLE_RULES.OFF: CyclomaticComplexity
-  private void validateStatement(
-      final KsqlEntityList entities, final String statementText, final Statement statement,
-      final Map<String, Object> streamsProperties) {
-    // CHECKSTYLE_RULES.ON: CyclomaticComplexity
-    if (statement == null) {
-      throw new KsqlRestException(
-          Errors.badStatement(
-              String.format("Unable to execute statement '%s'", statementText),
-              statementText, entities));
+  private List<PreparedStatement<?>> parseStatements(final String sql) {
+    try {
+      final List<PreparedStatement<?>> statements = ksqlEngine.parseStatements(sql);
+      checkPersistentQueryCapacity(statements, sql);
+      return statements;
+    } catch (final ParseFailedException e) {
+      throw new KsqlRestException(Errors.badStatement(e.getCause(), e.getSqlStatement()));
     }
+  }
 
-    if (Stream.of(
-        ListTopics.class, ListRegisteredTopics.class, ListStreams.class,
-        ListTables.class, ListQueries.class, ListProperties.class, RunScript.class,
-        ShowFunctions.class, DescribeFunction.class)
-        .anyMatch(c -> c.isInstance(statement))) {
-      return;
-    }
+  private void validateStatements(
+      final List<? extends PreparedStatement<?>> statements,
+      final Map<String, Object> propertyOverrides
+  ) {
+    final Map<Boolean, List<PreparedStatement<?>>> partitioned = statements.stream()
+        .collect(Collectors.groupingBy(stmt -> getCustomValidator(stmt) != null));
 
-    if (statement instanceof Query || statement instanceof PrintTopic) {
-      throw new KsqlRestException(Errors.queryEndpoint(statementText, entities));
-    }
+    partitioned
+        .getOrDefault(true, Collections.emptyList())
+        .forEach(stmt -> customValidateStatement(stmt, propertyOverrides));
 
-    if (statement instanceof ShowColumns) {
-      final ShowColumns showColumns = (ShowColumns) statement;
-      if (showColumns.isTopic()) {
-        describeTopic(statementText, showColumns.getTable().getSuffix());
-      } else {
-        describe(showColumns.getTable().getSuffix(), showColumns.isExtended());
+    final List<PreparedStatement<?>> standardValidated = partitioned
```

Comment:
Why skip? Because that's what the code did before. For example `TerminateQuery` has a custom validator, (which makes it skip validation).

Personally, I don't think `TerminateQuery` _should_ skip validation. However, I was planning on fixing that in a later PR.